Lib/test/test_shutil.py

```diff
@@ -579,6 +579,51 @@ def _filter(src, names):
             shutil.rmtree(src_dir)
             shutil.rmtree(os.path.dirname(dst_dir))
 
+    def test_copytree_arg_types_of_ignore(self):
+        join = os.path.join
+        exists = os.path.exists
+
+        src_dir = self.mkdtemp()
+        try:
+            os.mkdir(join(src_dir, 'test_dir'))
+            os.mkdir(os.path.join(src_dir, 'test_dir', 'subdir'))
+            write_file((src_dir, 'test_dir', 'subdir', 'test.txt'), '456')
+
+            def _ignore(src, names):
+                self.assertIsInstance(src, str)
+                self.assertIsInstance(names, list)
+                self.assertEqual(len(names), len(set(names)))
+                for name in names:
+                    self.assertIsInstance(name, str)
+                return []
+
+            try:
+                dst_dir = join(self.mkdtemp(), 'destination')
+                shutil.copytree(src_dir, dst_dir, ignore=_ignore)
+                self.assertTrue(exists(join(dst_dir, 'test_dir', 'subdir',
+                                            'test.txt')))
+            finally:
+                shutil.rmtree(dst_dir)
+
+            try:
+                dst_dir = join(self.mkdtemp(), 'destination')
+                shutil.copytree(pathlib.Path(src_dir), dst_dir, ignore=_ignore)
+                self.assertTrue(exists(join(dst_dir, 'test_dir', 'subdir',
+                                            'test.txt')))
+            finally:
+                shutil.rmtree(dst_dir)
+
+            try:
+                dst_dir = join(self.mkdtemp(), 'destination')
+                src_dir_entry = list(os.scandir(src_dir))[0]
+                self.assertIsInstance(src_dir_entry, os.DirEntry)
+                shutil.copytree(src_dir_entry, dst_dir, ignore=_ignore)
+                self.assertTrue(exists(join(dst_dir, 'subdir', 'test.txt')))
+            finally:
+                shutil.rmtree(dst_dir)
+        finally:
+            shutil.rmtree(src_dir)
```

Comment:
You can use `self.addCleanup(shutil.rmtree, path)` and avoid the try/finally blocks.

Reply:
Hmm, actually from the looks of it, that's already part of `self.mkdtemp`.

Reply:
I think the test should be ok now.